### ISSUE TITLE
MBQL: add :share clause

### ISSUE THREAD

**Issue by sbelak** [metabase/metabase#9431]
Currently our support of any kind of notion of percentages or distributions is severely lacking. This is more than just a presentation issue, as it precludes us from expressing concepts such as conversion rates (and comparisons of thereof which in turn would unlock MB for A/B test analysis use-case).

## Proposal

Add new aggregation clause to MBQL:

### :share
`[:share <filter-clause>]`

Returns the share of rows for which `filter-clause` is true.

## Implementation 

SQL CASE seems to be well supported https://modern-sql.com/feature/case and we only need the basic version.

## Open questions
*  is the name :share good or should it be something like :ratio or :percetage?

**Comment by salsakran** [comment]
I'm a huge fan of find MBQL primitives for CASE clauses. I see them all over the place in custom SQL, and it's one of the main reasons people are needing to fall back to SQL for relatively simple queries that imo should be doable in the GUI interface. 

I'm still thinking through the specifics of share/distribution. 

Can you give examples of distribution? I'm not sure I get exactly what you're proposing.

Also re: post-processing, I'm strongly in favor of getting out of the post-processing game wherever possible, and avoid digging the hole any deeper.

**Comment by sbelak** [comment]
You'd use distribution for things such as distribution of income by channel. Basically it's just a normalized count or sum. A strong argument could be made that it makes more sense for this to be a display setting since it doesn't introduce a structurally new aggregation. I'm leaning towards doing just :share first and then re-evaluate.

**Comment by salsakran** [comment]
I'm a strong 👍 for :share or something like it, and a strong 👎 for distribution.

For distribution, I'd much prefer composition to be more accessible (eg, dividing one query by another, preferably with some kind of query planning in the QB that squashes those down to a single query)

**Comment by camsaul** [comment]
@sbelak `CASE` is supported by all the SQL databases we support, we're using it in division situations

```sql
(x / (CASE WHEN y = 0 THEN NULL ELSE y END)
```

(SQL is ok with dividing by `NULL`, but not by zero)

Basically you just need to do

```clj
(hsql/call :case <condition> <arg> ...)
```

**Comment by sbelak** [comment]
After reading through a bunch of old issues, I'm pretty convinced that distribution is a visualization setting not an mbql thing. Retiring that part of the proposal in favour of #9433

One more question: Y'all happy with the name :share or should it be something like :ratio or :percentage

**Comment by sbelak** [comment]
Implemented in #9477 (closing this in favour of discussion in the PR)
